cmd/acra-keys/keys/command-line.go

```diff
@@ -0,0 +1,255 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Package keys defines reusable business logic of `acra-keys` utility.
+package keys
+
+import (
+	"errors"
+	"flag"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/cossacklabs/acra/cmd"
+	keystoreV1 "github.com/cossacklabs/acra/keystore"
+	filesystemV2 "github.com/cossacklabs/acra/keystore/v2/keystore/filesystem"
+	"github.com/cossacklabs/acra/logging"
+	"github.com/cossacklabs/acra/utils"
+	log "github.com/sirupsen/logrus"
+)
+
+// ServiceName constant for logging and configuration parsing.
+const ServiceName = "acra-keys"
+
+// DefaultKeyDirectory is the default path to key directory.
+const DefaultKeyDirectory = keystoreV1.DefaultKeyDirShort
+
+// DefaultConfigPath is the default path to service configuration file.
+var DefaultConfigPath = utils.GetConfigPathByName("acra-keys")
+
+// Sub-command names:
+const (
+	CmdReadKey    = "read"
+	CmdDestroyKey = "destroy"
+)
+
+// SupportedSubCommands lists supported sub-commands or CLI.
+var SupportedSubCommands = []string{
+	CmdReadKey,
+	CmdDestroyKey,
+}
+
+// Key kind constants:
+const (
+	KeyPoisonPublic   = "poison-public"
+	KeyPoisonPrivate  = "poison-private"
+	KeyStoragePublic  = "storage-public"
+	KeyStoragePrivate = "storage-private"
+	KeyZonePublic     = "zone-public"
+	KeyZonePrivate    = "zone-private"
+
+	KeyTransportConnector  = "transport-connector"
+	KeyTransportServer     = "transport-server"
+	KeyTransportTranslator = "transport-translator"
+)
+
+// Comman-line parsing errors:
+var (
+	ErrUnknownSubCommand = errors.New("unknown command")
+	ErrMissingKeyKind    = errors.New("missing key kind")
+	ErrMultipleKeyKinds  = errors.New("multiple key kinds")
+)
+
+// CommandLineParams describes all command-line options of acra-keys.
+type CommandLineParams struct {
+	KeyStoreVersion string
+	KeyDir          string
+	KeyDirPublic    string
+
+	ClientID string
+	ZoneID   string
+
+	Command string
+
+	ReadKeyKind    string
+	DestroyKeyKind string
+
+	readFlags    *flag.FlagSet
+	destroyFlags *flag.FlagSet
+}
+
+// Params provide global access to command-line parameters.
+var Params *CommandLineParams = &CommandLineParams{}
+
+// Register configures command-line parameter parsing.
+func (params *CommandLineParams) Register() {
+	flag.StringVar(&params.KeyStoreVersion, "keystore", "", "force key store format: v1 (current), v2 (new)")
+	flag.StringVar(&params.KeyDir, "keys_dir", DefaultKeyDirectory, "path to key directory")
+	flag.StringVar(&params.KeyDirPublic, "keys_dir_public", "", "path to key directory for public keys")
+	flag.StringVar(&params.ClientID, "client_id", "", "client ID for which to retrieve key")
```

Comment:
can we use all these global params separately for each command to provide <acra-keys> <command> <params> interface? copy them to each flagset. yes, we will duplicate some of them, but it is more common in use and we will not have hard to understand cli as discussed in "conversation" tab
golang has one popular library - https://github.com/spf13/cobra which do similar. we can look on some patterns and implementation details to use in acra.

Reply:
> can we use all these global params separately for each command to provide interface? 

I'm sorry, I don't follow your idea. Could you please elaborate? Do you mean to have a separate _complete_ flag set for each subcommand instead of splitting some global options into a root flag set and having subcommand-specific options in their own flag sets? Or have only a global flag set with all options of all subcommand, just some of them will not be used?

> golang has one popular library - https://github.com/spf13/cobra 

Honestly, I'd like to pull in something like that as a new dependency to make command line parsing more flexible. Using `flags` seems to be fine for flat configuration and an occasional option or two for a single-purpose tool. However, anything more complex becomes a mess.

What do you think about adding it as a dependency?
(In a separate PR, only for acra-keys at first.)

Reply:
> I'm sorry, I don't follow your idea. Could you please elaborate? Do you mean to have a separate complete flag set for each subcommand instead of splitting some global options into a root flag set and having subcommand-specific options in their own flag sets? Or have only a global flag set with all options of all subcommand, just some of them will not be used?

exclude `--client_id`/`--zone_id` from global options and require them per each command. to avoid cases with `--client_id=client <command> --client_id=client2`

> What do you think about adding it as a dependency?
> (In a separate PR, only for acra-keys at first.)

actually personally I didn't do it because I wasn't sure that this library ok for our license politic) because this library uses too many other third-parties with different licenses) so I haven't anything against it if it's okay with license.

Reply:
> because this library uses too many other third-parties with different licenses

Oh... For some reason I thought it's some minimal argument parser or something. However, after I did

```
$ cat go.mod
module example.com/module

go 1.13

require github.com/spf13/cobra v1.0.0
```
and then
```
$ cat go.sum | awk '{print $1}' | sort -u | wc -l
      98
```
<details>
<summary>Dependency list</summary>

```
$ cat go.sum | awk '{print $1}' | sort -u
cloud.google.com/go
github.com/BurntSushi/toml
github.com/OneOfOne/xxhash
github.com/alecthomas/template
github.com/alecthomas/units
github.com/armon/consul-api
github.com/beorn7/perks
github.com/cespare/xxhash
github.com/client9/misspell
github.com/coreos/bbolt
github.com/coreos/etcd
github.com/coreos/go-semver
github.com/coreos/go-systemd
github.com/coreos/pkg
github.com/cpuguy83/go-md2man/v2
github.com/davecgh/go-spew
github.com/dgrijalva/jwt-go
github.com/dgryski/go-sip13
github.com/fsnotify/fsnotify
github.com/ghodss/yaml
github.com/go-kit/kit
github.com/go-logfmt/logfmt
github.com/go-stack/stack
github.com/gogo/protobuf
github.com/golang/glog
github.com/golang/groupcache
github.com/golang/mock
github.com/golang/protobuf
github.com/google/btree
github.com/google/go-cmp
github.com/gorilla/websocket
github.com/grpc-ecosystem/go-grpc-middleware
github.com/grpc-ecosystem/go-grpc-prometheus
github.com/grpc-ecosystem/grpc-gateway
github.com/hashicorp/hcl
github.com/inconshreveable/mousetrap
github.com/jonboulle/clockwork
github.com/julienschmidt/httprouter
github.com/kisielk/errcheck
github.com/kisielk/gotool
github.com/konsorten/go-windows-terminal-sequences
github.com/kr/logfmt
github.com/kr/pretty
github.com/kr/pty
github.com/kr/text
github.com/magiconair/properties
github.com/matttproud/golang_protobuf_extensions
github.com/mitchellh/go-homedir
github.com/mitchellh/mapstructure
github.com/mwitkow/go-conntrack
github.com/oklog/ulid
github.com/pelletier/go-toml
github.com/pkg/errors
github.com/pmezard/go-difflib
github.com/prometheus/client_golang
github.com/prometheus/client_model
github.com/prometheus/common
github.com/prometheus/procfs
github.com/prometheus/tsdb
github.com/rogpeppe/fastuuid
github.com/russross/blackfriday/v2
github.com/shurcooL/sanitized_anchor_name
github.com/sirupsen/logrus
github.com/soheilhy/cmux
github.com/spaolacci/murmur3
github.com/spf13/afero
github.com/spf13/cast
github.com/spf13/cobra
github.com/spf13/jwalterweatherman
github.com/spf13/pflag
github.com/spf13/viper
github.com/stretchr/objx
github.com/stretchr/testify
github.com/tmc/grpc-websocket-proxy
github.com/ugorji/go
github.com/xiang90/probing
github.com/xordataexchange/crypt
go.etcd.io/bbolt
go.uber.org/atomic
go.uber.org/multierr
go.uber.org/zap
golang.org/x/crypto
golang.org/x/lint
golang.org/x/net
golang.org/x/oauth2
golang.org/x/sync
golang.org/x/sys
golang.org/x/text
golang.org/x/time
golang.org/x/tools
google.golang.org/appengine
google.golang.org/genproto
google.golang.org/grpc
gopkg.in/alecthomas/kingpin.v2
gopkg.in/check.v1
gopkg.in/resty.v1
gopkg.in/yaml.v2
honnef.co/go/tools
```

</details>

In this case I'd say, “No, thanks”, _regardless_ of whether the licenses of all these libraries are acceptable or not. I don't think we should multiply our indirect dependency list by 10 because of that library.